File: clients/dart/ttid.dart
// TTID client — drives the `ttid` binary's persistent NDJSON loop.
//
// Stdlib only (dart:io, dart:convert). Requires the `ttid` binary on PATH or an
// explicit path. One long-lived subprocess.
//
//   final t = await Ttid.open();
//   final id = await t.generate();               // new id
//   final up = await t.generate(id as String);   // advance it
//   await t.generate(up as String, true);        // mark deleted
//   print(await t.decodeTime(up));               // {createdAt: ..., updatedAt: ...}
//   print(await t.isTtid(id));                   // {valid: true, createdAt: ...}
//   print(await t.isUuid('not-a-uuid'));         // {valid: false}
//   await t.close();
//
// Each method builds the request and returns the op's `result` (throwing on
// failure). Method names follow Dart's lowerCamelCase (acronyms as words).
// request(op) is the raw escape hatch returning the full response map.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

class TtidException implements Exception {
  final String message;
  TtidException(this.message);
  @override
  String toString() => 'TtidException: $message';
}

class Ttid {
  final Process _proc;
  final StreamIterator<String> _lines;
  final StreamSubscription<List<int>> _stderr;

  Ttid._(this._proc, this._lines, this._stderr);

  /// Start a warm ttid process. [binary] defaults to "ttid".
  static Future<Ttid> open([String binary = 'ttid']) async {
    final proc = await Process.start(binary, ['exec', '--loop']);
    final lines = StreamIterator(
        proc.stdout.transform(utf8.decoder).transform(const LineSplitter()));
    // `Process.start` pipes all three streams and Dart offers no per-stream
    // mode, so stderr has to be handled here or it goes nowhere. Forwarding it
    // matches the other clients, which inherit the parent's stderr, and stops a
    // chatty child from filling the pipe buffer and blocking on a write.
    final stderrSubscription = proc.stderr.listen(stderr.add);
    return Ttid._(proc, lines, stderrSubscription);
  }

  /// Send one raw machine-protocol op; return the full response map.
  Future<Map<String, dynamic>> request(Map<String, dynamic> op) async {
    _proc.stdin.writeln(jsonEncode(op));
    await _proc.stdin.flush();
    // ponytail: one call in flight; await each call before the next, or guard
    // with a lock if you pipeline.
    if (!await _lines.moveNext()) throw TtidException('ttid closed the stream');
    return jsonDecode(_lines.current) as Map<String, dynamic>;
  }

  Future<dynamic> _op(String op, Map<String, dynamic> fields) async {
    final payload = <String, dynamic>{'op': op};
    fields.forEach((key, value) {
      if (value != null) payload[key] = value;
    });
    final response = await request(payload);
    if (response['ok'] != true) {
      final error = response['error'];
      throw TtidException(
          (error is Map && error['message'] is String) ? error['message'] : 'ttid error');
    }
    return response['result'];
  }

  /// Generate a new TTID, or advance [id] ([delete] to tombstone). Pass null for a fresh id.
  Future<dynamic> generate([String? id, bool delete = false]) =>
      _op('generate', {'id': id, 'delete': delete ? true : null});

  Future<dynamic> decodeTime(String id) => _op('decodeTime', {'id': id});
  Future<dynamic> isTtid(String id) => _op('isTTID', {'id': id});
  Future<dynamic> isUuid(String id) => _op('isUUID', {'id': id});

  /// Close stdin so the loop ends, wait for the process to exit, and release
  /// both pipes.
  ///
  /// **`_lines.cancel()` is load-bearing.** A `StreamIterator` holds a live
  /// subscription on the child's stdout, and that subscription keeps the Dart
  /// event loop running after `main` returns — a program using this client and
  /// spawned without a TTY would never exit. Verified by removing each cancel
  /// in turn: dropping this one reproduces the hang, dropping the stderr one
  /// does not.
  ///
  /// Safe to call more than once.
  Future<void> close() async {
    await _proc.stdin.close();
    await _proc.exitCode;
    await _lines.cancel();
    await _stderr.cancel();
  }
}

File: clients/go/ttid.go
// Package ttid drives the `ttid` binary's persistent NDJSON loop.
//
// Stdlib only. Requires the `ttid` binary on PATH or an explicit path. One
// long-lived subprocess.
//
//	t, _ := ttid.Open("ttid")
//	defer t.Close()
//	id, _ := t.Generate("", false)      // new id
//	up, _ := t.Generate(id.(string), false)
//	times, _ := t.DecodeTime(up.(string))
//	valid, _ := t.IsTTID(id.(string))
//	uuid, _ := t.IsUUID("...")
//
// Each method builds the request and returns the op's `result` (or an error on
// failure). Method names mirror the machine-protocol ops in Go's PascalCase.
// Request(op) is a raw escape hatch returning the full response.
package ttid

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os/exec"
	"sync"
)

type TTID struct {
	cmd  *exec.Cmd
	pipe io.WriteCloser
	in   *bufio.Writer
	out  *bufio.Reader
	mu   sync.Mutex
}

// Open starts a warm ttid process. binary defaults to "ttid".
func Open(binary string) (*TTID, error) {
	if binary == "" {
		binary = "ttid"
	}
	cmd := exec.Command(binary, "exec", "--loop")
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	if err := cmd.Start(); err != nil {
		return nil, err
	}
	return &TTID{cmd: cmd, pipe: stdin, in: bufio.NewWriter(stdin), out: bufio.NewReader(stdout)}, nil
}

// Request sends one raw machine-protocol op and returns the full response.
func (t *TTID) Request(op map[string]any) (map[string]any, error) {
	t.mu.Lock() // ponytail: one call in flight; drop the lock only if you pipeline
	defer t.mu.Unlock()
	line, err := json.Marshal(op)
	if err != nil {
		return nil, err
	}
	if _, err := t.in.Write(append(line, '\n')); err != nil {
		return nil, err
	}
	if err := t.in.Flush(); err != nil {
		return nil, err
	}
	reply, err := t.out.ReadBytes('\n')
	if err != nil {
		return nil, fmt.Errorf("ttid closed the stream: %w", err)
	}
	var resp map[string]any
	if err := json.Unmarshal(reply, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (t *TTID) op(name string, fields map[string]any) (any, error) {
	payload := map[string]any{"op": name}
	for k, v := range fields {
		if v != nil {
			payload[k] = v
		}
	}
	resp, err := t.Request(payload)
	if err != nil {
		return nil, err
	}
	if ok, _ := resp["ok"].(bool); !ok {
		msg := "ttid error"
		if e, ok := resp["error"].(map[string]any); ok {
			if m, ok := e["message"].(string); ok {
				msg = m
			}
		}
		return nil, fmt.Errorf("%s", msg)
	}
	return resp["result"], nil
}

// Generate creates a new TTID, or advances id (del=true to tombstone). Pass "" for a fresh id.
func (t *TTID) Generate(id string, del bool) (any, error) {
	fields := map[string]any{}
	if id != "" {
		fields["id"] = id
	}
	if del {
		fields["delete"] = true
	}
	return t.op("generate", fields)
}

func (t *TTID) DecodeTime(id string) (any, error) {
	return t.op("decodeTime", map[string]any{"id": id})
}

func (t *TTID) IsTTID(id string) (any, error) {
	return t.op("isTTID", map[string]any{"id": id})
}

func (t *TTID) IsUUID(id string) (any, error) {
	return t.op("isUUID", map[string]any{"id": id})
}

// Close ends the loop and waits for the process to exit.
func (t *TTID) Close() error {
	t.in.Flush()
	t.pipe.Close()
	return t.cmd.Wait()
}
